tests/functional/lib/receptorcontrol/receptorcontrol.go

```diff
@@ -158,3 +165,152 @@ func (r *ReceptorControl) Status() (*netceptor.Status, error) {
 
 	return &status, nil
 }
+
+// WorkSubmit begins work on remote node
+func (r *ReceptorControl) WorkSubmit(node, serviceName string) (string, error) {
+	_, err := r.WriteStr(fmt.Sprintf("work submit %s %s\n", node, serviceName))
+	if err != nil {
+		return "", err
+	}
+	r.ReadStr()
+	r.CloseWrite()
+	response, err := r.ReadAndParseJSON()
+	if err != nil {
+		return "", err
+	}
+	unitID := fmt.Sprintf("%v", response["unitid"])
+	return unitID, nil
+}
+
+// WorkStart begins work on local node
+func (r *ReceptorControl) WorkStart(workID string) (string, error) {
+	_, err := r.WriteStr(fmt.Sprintf("work start %s\n", workID))
+	if err != nil {
+		return "", err
+	}
+	_, err = r.ReadStr() // flush response
+	if err != nil {
+		return "", err
+	}
+	r.CloseWrite() // clost write half to signal EOF
+	response, err := r.ReadAndParseJSON()
+	if err != nil {
+		return "", err
+	}
+	r.Close()
```

Comment:
Im not sure i understand why we're doing `r.Close()` here

Reply:
work submit expects stdin (until EOF) from the socket. A few lines above I close the Write side of the socket (`r.CloseWrite()`), which triggers the end of stdin.

But now that the controller write is closed, the controller is worthless for further use, so I just close the rest of the controller. 

I'd really like to find a better way to handle this, but I am not sure there is one.

Reply:
ah i saw `WorkSubmit` closes the write side but not the whole socket so i assumed the other test was wrong, but it sounds like `WorkSubmit` is the one that was wrong

Reply:
Strictly speaking, with a TCP socket, you only _can_ close the write side of the socket, by sending a FIN packet.  Whether the peer decides to send you more bytes appear is not under your control - though of course you can decide to ignore them.

The idea with work submission is that the submitter sends the stdin and then half-closes the socket, which lets the receiver know that this is the end of the data.  The receiver then sends a JSON response including the unit ID, and closes its side.  If you don't care about receiving the JSON, you can go ahead and fully close the socket, which just means the JSON response gets discarded.